### PR TITLE
Fix for replication bug with empty store

### PR DESCRIPTION
**Pull request by lightningrob**
Fixes issue #916

### CODE REVIEW

File: ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java [hunk 1]
```diff
@@ -622,7 +622,7 @@ private void readFromFileAndPersistIfNecessary(String mountPath) throws Replicat
                       && remoteReplicaInfo.getReplicaId().getReplicaPath().equals(replicaPath)) {
                     logger.info("Read token for partition {} remote host {} port {} token {}", partitionId, hostname,
                         port, token);
-                    if (partitionInfo.getStore().getSizeInBytes() > 0) {
+                    if (partitionInfo.getStore().isEmpty() == false) {
```
Comment (vgkholla) [inline ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java]: minor: `if (!partitionInfo.getStore().isEmpty())`

File: ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java [hunk 2]
```diff
@@ -1278,14 +1278,19 @@ Offset getCurrentEndOffset() {
...
    * @param indexSegments the map of index segment start {@link Offset} to {@link IndexSegment} instances
    */
   private Offset getCurrentEndOffset(ConcurrentSkipListMap<Offset, IndexSegment> indexSegments) {
+    // If no indexSegments exist, return start offset of first log segment
     return indexSegments.size() == 0 ? getStartOffset(indexSegments)
         : indexSegments.lastEntry().getValue().getEndOffset();
   }
 
+  boolean isEmpty() {
+    return (validIndexSegments.isEmpty());
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java]: minor: braces unnecessary

Comment (vgkholla) [reply ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java]: Also add javadoc like the other functions
```
/**
* @return true if the index contains no segments
*/
```

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java [hunk 3]
```diff
@@ -983,6 +983,7 @@ private void setupTestState() throws InterruptedException, StoreException {
     // being picked for delete.
     time.sleep(Time.MsPerSec);
     long expectedStoreSize;
+    assertTrue("Expected empty store", store.isEmpty());
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: you need to add a similar check in `CuratedLogIndexState`

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java [hunk 4]
```diff
@@ -992,6 +993,7 @@ private void setupTestState() throws InterruptedException, StoreException {
       // first log segment is filled to capacity.
       addCuratedData(segmentCapacity);
       assertEquals("Store size not as expected", expectedStoreSize, store.getSizeInBytes());
+      assertFalse("Expected nonempty store", store.isEmpty());
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: this check is fine here but also required as soon as the first put occurs in `addCuratedData()` (to check that it returns `true` as soon as the first put arrives). 

Same with `CuratedLogIndexState`

File: ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java (hunk 1 above)
Comment (lightningrob) [reply ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java]: I realize that's the standard and I'll change it.  I do feel that the ! is easy to miss and an explicit false check is clearer.  That's why I prefer it.

File: ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java [hunk 5]
```diff
@@ -987,6 +987,9 @@ public long getSizeInBytes() {
         return 0;
       }
 
+      @Override
+      public boolean isEmpty() { return false; }
```
Comment (pnarayanan) [inline ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: You might want to follow these steps for adopting the ambry codestyle: https://github.com/linkedin/ambry/wiki/Development%20Environment

Comment (lightningrob) [reply ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: Will do

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java (hunk 3 above)
Comment (lightningrob) [reply ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: Done.

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java (hunk 4 above)
Comment (lightningrob) [reply ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: Done.

**Comment by pnarayanan** [comment]
I wish we had a test for replication token reset logic that would have caught the issue this patch addresses (and one that would have succeeded with this change). I understand that it might not be a quick test to add, but I think it is worth adding in a future PR.

**Pull request merged**